hypothesis-python/src/hypothesis/internal/reflection.py

```diff
@@ -219,6 +220,21 @@ def ast_arguments_matches_signature(args, sig):
     return expected == [(p.name, p.kind) for p in sig.parameters.values()]
 
 
+def is_func_param_called_within(f, name):
+    """Is the given name referenced within f?"""
+    try:
+        tree = ast.parse(textwrap.dedent(inspect.getsource(f)))
+    except (OSError, ValueError):
+        return True
+    else:
+        return any(
+            isinstance(node, ast.Name)
+            and node.id == name
+            and isinstance(node.ctx, ast.Load)
```

Comment:
Can we expand this to check for a [`Call` node where the `func` attribute](https://greentreesnakes.readthedocs.io/en/latest/nodes.html#Call) matches this condition?

Reply:
We thought long and hard about this one. What if someone calls `draw` in some dynamic way that isn't a direct function call?
For example, quite insane, but `draw.__call__(...)` would still be valid.
An even more extreme case would be the code `eval("draw")(...)` which doesn't even reference the name `draw` but would work....

Reply:
Then they will find that their insane calling convention is deprecated, and they have to change it (or add an `if False: draw()`)!

This is a very good argument against making this an error immediately, but deprecation warnings are OK 🙂

Reply:
My two examples are weird, but show that there are ways of calling `draw` indirectly.
You are the maintainer, so I am assuming directly checking for a function call is sensible.

Reply:
I changed the check to look for an explicit function call. It works great but now breaks the [test `test_can_use_pure_args` from `tests/cover/test_composite.py`](https://github.com/HypothesisWorks/hypothesis/blob/eed37f681b2baf007034a453193d8e314c3c15f4/hypothesis-python/tests/cover/test_composite.py#L78):

https://github.com/HypothesisWorks/hypothesis/blob/eed37f681b2baf007034a453193d8e314c3c15f4/hypothesis-python/tests/cover/test_composite.py#L78-L83

I'm assuming that “using pure args” is now deprecated behaviour and should be removed from the tests?

Reply:
Um.  Let's skip the new check in the case that `params = get_signature(f).parameters`; `if params and list(params)[0].kind is inspect.Parameter.VAR_POSITIONAL: <skip this check>`

Messy, but I guess we're committed to this working 😅

Reply:
Understood 👌